Fix memory leak on mesh collider destroy

MushAsterion opened the pull request:
Fixes #4267 

I know it doesn't match the suggestion from the issue. However I think having a `CollisionSystemImpl#destroyShape` function sound more beneficial than making the parent class aware of an extension. It also removes override of the `CollisionMeshSystemImpl#remove` function and make sure there is an actual shape to destroy.

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

willeastcott commented:
Hey @LeXXik - you submitted the original issue. Got any thoughts here? 😄

LeXXik commented:
@willeastcott so, are we keeping the data object on the collision component then? I thought you might have been cooking an update for it, hence the last one didn't go through. 

@MushAsterion, I think it would be good if we add a test that covers this change. If you are short on time, I can add it later this week. The test would need to cover at least clearing of the collision shapes (primitive, compounds, mesh and triggers). We don't need to test Ammo, so its methods can be stubbed. We just need to make sure we are calling the destroy methods on our side whenever needed.

MushAsterion commented:
> I think it would be good if we add a test that covers this change. If you are short on time, I can add it later this week. The test would need to cover at least clearing of the collision shapes (primitive, compounds, mesh and triggers). We don't need to test Ammo, so its methods can be stubbed. We just need to make sure we are calling the destroy methods on our side whenever needed.

That makes sense and time is not really a problem to me. However I have to admit that I have absolutely no knowledge or skills on this matter and when I look at the existing tests none already handle Ammo which sound quite complex to mirror especially on complex shapes... I would love to learn so if you have any example I can use to do so I'll be more than happy to give it a try!

LeXXik commented:
@MushAsterion you can take a look at these tests that handle Ammo, as an example:

https://github.com/playcanvas/engine/pull/4144/commits/2a07dacae527eb3abcabc0efafd884bd969398cc

MushAsterion commented:
Thank you @LeXXik! Took a bit of time but I managed to mirror Ammo based on the IDL from playcanvas/ammo.js repo and added the tests for destruction. Did my best to keep it simple but it might be still a bit messy...

I wasn't sure about creating a `.mjs` file only for Ammo but this way it keeps it separate since it is quite long. I know not all functions are used by the engine at the moment but as I was on it I just added everything based off the IDL so it won't cause issue if more features are used.

LeXXik commented on test/ammo.mjs:
```diff
@@ -0,0 +1,1635 @@
+
+// Stub Ammo
+
+// Note: do not use arrow functions in the Ammo root object, since the engine will
```
This comment is not applicable here and can be removed.

MushAsterion replied:
It's removed!

LeXXik reviewed (approved):
This is amazing. I'd say you went a bit overkill with the Ammo stub module :) However, as you mentioned it can definitely be used to cover all current and potential future features we add to the engine.

mvaligursky commented on src/framework/components/collision/system.js:
```diff
@@ -155,6 +154,15 @@ class CollisionSystemImpl {
         }
     }
 
+    destroyShape(data) {
+        if (!data.shape) {
+            return;
+        }
+
+        Ammo.destroy(data.shape);
+        data.shape = null;
+    }
```
```suggestion
    destroyShape(data) {
        if (data.shape) {
            Ammo.destroy(data.shape);
            data.shape = null;
        }
    }
```

MushAsterion replied:
In fact this was to follow contributing guidelines on [early exiting](https://github.com/playcanvas/engine/blob/main/.github/CONTRIBUTING.md#exit-logic-early), are you sure this would really better?

mvaligursky replied:
I understand where you're coming from, but for a short function like this, the early exit provides no advantages.

MushAsterion replied:
Alright, it's committed then

mvaligursky reviewed (approved):
The PR looks amazing. I added a small suggestion to the code, to make it slightly more compact.

The test looks a lot better than expected, I'm happy here.

Approving, but would like @willeastcott's thoughts here as well.

mvaligursky commented:
Perhaps you can then split this PR into two .. one which fixes the memory leak, and the test as a separate. This would allow us to merge the fix while waiting for latest Ammo.

MushAsterion commented:
Moved the tests to #5322!

mvaligursky reviewed (approved):
Perfect, thanks!
